src/crimson/os/seastore/segment_cleaner.h

```diff
@@ -44,6 +89,199 @@ struct segment_info_t {
   bool is_open() const {
     return state == Segment::segment_state_t::OPEN;
   }
+  void set_unavail() {
+    usii = smit->usis.emplace(smit->usis.end(), this);
+  }
+};
+
+class SegmentCleaner;
+
+// this segment_info_set_t class is intend to avoid iterating a device map
+// for a specific device, which could be expensive if we do it for every io
+template <typename SegmentInfoT>
+class segment_info_set_t {
```

Comment:
It seems like this could just use a std::map<device_id_t, std::vector<std::optional<SegmentInfoT>> internally.  I don't see  a justification for the complexity here.  If this is necessary, I'd expect the comment above to specifically explain why a simpler structure won't work as well as a concise and complete explanation of the internal invariants.

Reply:
Actually, I thought about using std::map<device_id_t, std::vector<std::optional>>. However, I thought if we use the top 4 bits of segment id as the device id, which means we may have up to 16 devices, and `segment_info_set_t::operator[]` is called for every journal writes when we do `SegmentCleaner::mark_space_used` or `SegmentCleaner::mark_space_free`, so I was worried that doing a 16-element search for every transaction is a bit expensive. Honestly, I was not sure if I was considering this right.

Reply:
Without evidence that it actually makes a difference, this complexity isn't worthwhile.  Please swap to the simpler implementation.

Reply:
Also, std::vector<std::vector<std::optional>> would still be simpler.

Reply:
Got it, will do:-)

Reply:
@athanatos Um, it seems that except for the perf reason, we also need a place to keep each segment manager's empty segments, opened segments and other similar stuff so that we can keep track of things like each segment manager's space usage. So I think maybe it's worthwhile to keep segment_info_set_t as a structure to hold that information. I'm changing segment_info_set_t as follows:

```
struct segment_manager_info_t {
    std::vector<SegmentInfoT> segment_infos
};

struct segment_info_set_t {
    std::vector<std::optional<segment_manager_info_t>> sm_info_vec;
};
```

What do you think? Thanks.

Reply:
If I'm understanding that correctly, that seems reasonable.  Primarily, I find the segment_offset indirection unnecessarily complex.  As long we avoid that by embedding a seperate segment_infos vector for each manager, it should be simple enough.

Reply:
Got it:-)